swim/heal_partition.go

```diff
@@ -0,0 +1,127 @@
+// suspect. For members that are faulty in partition B, we fabricate updates with
+// a suspect in stead of faulty status and disseminate them through A. We also do
+// the reverse. We then create a connection from partition A to B and visa versa.
+// We do this by reincarnating this node and sending the respective update to the
+// target. The target has also been reincarnated due to previous actions, this
+// update is now applied on this node's membership to create the connection from
+// partition B to partition A.
+//
+// The heal alogrithm in more detail:
+// 1. send join request to the target, this tells us whether a partition exists
+// 2. collect changes for partition B:
+//    - collect all faulty members of this membership,
+//      now change the status of the updates from faulty to suspect;
+//    - reincarnate yourself and collect the respective change
+// 3. send the changes to the target, the changes will be disseminated
+//    through B automatically
+// 4. collect changes for partition A:
+//    - collect all faulty members from the join response of step 1,
+//      now change the statues of the update from faulty to suspect;
+//    - t has reincarnated itself upon receiving the ping from step 3,
+//      find that update from the ping response. If t has reincarnated
+//      for another reason the change might not be present. In this case
+//      we search for the change in the join response of step 1.
+// 5. apply the changes for partition A to the nodes membership,
+//    the changes will be disseminated through A automatically
+func AttemptHeal(node *Node, target string) error {
+	// If join request succeeds a partition is detected,
+	// this node will now coordinate the healing mechanism.
+	joinRes, err := sendJoinRequest(node, target, time.Second/2)
+	if err != nil {
+		return err
+	}
+
+	// Create changes for partition B. Partition B is the partition the targets node is part of.
+	csForB := node.disseminator.FullSync()
+	// discard changes that are not faulty and change faulty to suspect
+	for i := 0; i < len(csForB); i++ {
+		// TODO (wieger): make filtering more precice
+		if csForB[i].Status != Faulty {
+			csForB[i] = csForB[len(csForB)-1]
+			csForB = csForB[:len(csForB)-1]
+			i--
+		} else {
+			csForB[i].Status = Suspect
+		}
+	}
+
+	// Bump inc no of this node and add that change for B.
+	bumpIncNoChange := node.memberlist.Reincarnate()
+	csForB = append(csForB, bumpIncNoChange...)
+
+	// Ping b with the changes that are just assembled.
+	res, err := sendPingWithChanges(node, target, csForB, time.Second)
+	if err != nil {
+		return err
+	}
+
+	// Create changes for partition A. Partition A is the partition this node is part of.
+	csForA := joinRes.Membership
+	// discard changes that are not faulty and change faulty to suspect
+	for i := 0; i < len(csForA); i++ {
+		if csForA[i].Status != Faulty {
+			csForA[i] = csForA[len(csForA)-1]
```

Comment:
You are removing an element from a slice in multiple places. Maybe it makes sense to have a small method for it?

Not for DRY, but because I would find it easier to read this:

```golang
csForA = remove_item(csForA, i)
```
rather than
```golang
csForA[i] = csForA[len(csForA)-1]
csForA = csForA[:len(csForA)-1]
```

I know it's only two lines, but they would make it easier to scan the business logic.